core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -0,0 +1,145 @@
+    private final Subscriber<? super T> delegate;
+    private final EventExecutor executor;
+    private final StreamTimeoutMode timeoutMode;
+    private final Duration timeoutDuration;
+    private final long timeoutNanos;
+    private ScheduledFuture<?> timeoutFuture;
+    private Subscription subscription;
+    private long lastEventTimeNanos;
+    private boolean isTerminated;
+
+    TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                      StreamTimeoutMode timeoutMode) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.executor = requireNonNull(executor, "executor");
+        this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        timeoutNanos = timeoutDuration.toNanos();
+        this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+    }
+
+    private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+        return executor.schedule(this, delay, unit);
+    }
+
+    private boolean attemptTerminate() {
+        if (isTerminated) {
+            return false;
+        }
+        isTerminated = true;
+        return true;
+    }
+
+    public void cancelSchedule() {
+        if (!timeoutFuture.isCancelled()) {
+            timeoutFuture.cancel(false);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+            final long currentTimeNanos = System.nanoTime();
+            final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+            if (elapsedNanos < timeoutNanos) {
+                final long delayNanos = timeoutNanos - elapsedNanos;
+                timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
+                return;
+            }
+        }
+        if (!attemptTerminate()) {
+            return;
+        }
+        delegate.onError(new TimeoutException(
+                String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
+        subscription.cancel();
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        subscription = s;
+        lastEventTimeNanos = System.nanoTime();
+        timeoutFuture = scheduleTimeout(timeoutNanos, TimeUnit.NANOSECONDS);
+        delegate.onSubscribe(s);
+    }
+
+    @Override
+    public void onNext(T t) {
+        if (isTerminated || timeoutFuture.isCancelled()) {
+            return;
+        }
+        switch (timeoutMode) {
+            case UNTIL_NEXT:
+                lastEventTimeNanos = System.nanoTime();
+                break;
+            case UNTIL_FIRST:
+                timeoutFuture.cancel(false);
+                break;
+            case UNTIL_EOS:
+                break;
+        }
+        delegate.onNext(t);
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (!attemptTerminate()) {
+            return;
+        }
```

Comment:
How about just inlining here and in onComplete?
```suggestion
        if (completed) {
            return;
        }
        completed = true;
```